### ISSUE TITLE
[OJS 3.1] Migration from 2: multiple reviewer uploaded files collapsed into one

### ISSUE THREAD

**Issue by pkel** [pkp/pkp-lib#3789]
When migrating from OJS 2.4.8 to OJS 3.1.1-1, the following problem occurs.

In the old system, reviewers were able to upload multiple files together with their review. They could decide whether the files are meant for the author or not. (Reviewer name erased)

![ojs2-reviewer-uploaded-files](https://user-images.githubusercontent.com/11337857/41403007-127a2102-6fc4-11e8-8ff1-109612ae51b2.png)

After migration to version 3, only one file is left.

![ojs-reviewer-uploaded-files](https://user-images.githubusercontent.com/11337857/41402658-2cebfcdc-6fc3-11e8-8d83-33b7ee5385bc.png)

Additionally, this files history, accessed via the *more information* button, does not load. It simply shows a progress sign.

![loading-revision-file-history](https://user-images.githubusercontent.com/11337857/41402759-77be63da-6fc3-11e8-96cf-ad679a361d8d.png)

~Chromium print the following error on the console:~ (caused by adblock)

```
Failed to load resource: net::ERR_BLOCKED_BY_CLIENT
https://dev0.jstatsoft.org/$$$call$$$/grid/event-log/submission-file-event-log-grid/fetch-grid?submissionId=2672&fileId=6506&stageId=3&_=1528965841656
```
The presented history is empty. There is no information about the second reviewer file.

![history_empty](https://user-images.githubusercontent.com/11337857/41403733-dce6b8f0-6fc5-11e8-800e-9daa18c4e5b6.png)

**Comment by pkel** [comment]
I see know, that the history is empty for all migrated files. This is worth another issue.

**Comment by retostauffer** [comment]
Same OJS3 instance as @pkel adresses. This is what the database knows about the "Reviewer Files" (`file_id = 6506`; see screenshot included by @pkel):

* `select * from submission_files where file_id = 6506;`

```
+---------+----------+----------------+-----------------+---------------+-----------------+-----------+-------------------------------+------------+----------+---------------------+---------------------+--------
| file_id | revision | source_file_id | source_revision | submission_id | file_type       | file_size | original_file_name            | file_stage | viewable | date_uploaded       | date_modified       | assoc_i
+---------+----------+----------------+-----------------+---------------+-----------------+-----------+-------------------------------+------------+----------+---------------------+---------------------+--------
|    6506 |        1 |           NULL |            NULL |          2672 | application/pdf |     41372 | CoverLetter_Review of mgm.pdf |          5 |        0 | 2016-09-01 21:53:41 | 2016-09-01 21:53:41 |     163
|    6506 |        2 |           NULL |            NULL |          2672 | application/pdf |     55592 | Comments_mgm.pdf              |          5 |        1 | 2016-09-01 21:53:58 | 2016-09-01 21:53:58 |     163
+---------+----------+----------------+-----------------+---------------+-----------------+-----------+-------------------------------+------------+----------+---------------------+---------------------+--------
```

These are two different files uploaded by one of our reviewers. OJS3 (after migration) thinks that the second file is a revised version of the first one.

**Comment by bozana** [comment]
Could you take a look at these two issues: https://github.com/pkp/pkp-lib/issues/3510 and https://github.com/pkp/pkp-lib/issues/3502 -- is this the same problem i.e. issue? I.e. different reviewer files were actually the revisions of the same file, which they are in OJS 3 as well, just that in OJS 3 one can only access the latest file revision, which is a general problem for all files grids and will be considered with this issue: https://github.com/pkp/pkp-lib/issues/2130...

**Comment by retostauffer** [comment]
Dear @bozana . Thanks for pointing that out and sorry we've missed it! It is kind of the same issue, however, in our case the two files are not revisions of the same file (at least not in all cases), but e.g., a _cover letter_ and the _review itself_ which should both be visible (one for the author, the other one for the editors). Easy to misinterpret in OJS2 as it looks as being two different files, however, the migration on the database level is correct as the files are also revised versions in OJS2. In #3502 I would have voted for option two ;) but I think it should not be a big problem to write a custom patch for our journal.

And thanks to pointing out #2130, I think it is crucial to create the `event_log` entries for all existing files such that the file-history for old files can be accessed via the interface.

So we (@pkel) might close this issue at this point.

**Comment by pkel** [comment]
This is similar to / same as #3502. But I do not think the problem is fixed well. The v2 interface clearly allows to upload multiple files as reviewer.

Before "revision":
![1](https://user-images.githubusercontent.com/11337857/41425631-9b97a472-7001-11e8-9e9c-45412424da8b.png)

After "revision":
![two](https://user-images.githubusercontent.com/11337857/41425637-9e0a26d0-7001-11e8-9bdb-5150c4f0c36b.png)

After the upload, they are visible as separate files without any notion of revision.  While the v2 database stores the files as revisions, the average user will not notice this. The fact that visibility options could be set separately undermines the impression of two separate files.

The v3 interface allows reviewers to upload multiple files again. Now, they are stored separately in the in the database.

From a users perspective the behavior did not change. The representation in the database changed. The upgrade script does not map the old usage pattern to the new user interface. In my opinion, the v2 revisions of reviewer files should be migrated as separate files into the v3 database.

**Comment by retostauffer** [comment]
I agree with @pkel. Extra check: seems that we have, in total, 18 files with assoc_type 517 and a revision `>1` (**migrated database!**, each `file_id` counted only once;).

`select assoc_type, count(*) from submission_files where assoc_type is not null and revision > 1 group by assoc_type,revision;` 

+------------+----------+
| assoc_type | count(*) |
+------------+----------+
|        517 |       18 |
|        517 |        3 |
|        517 |        1 |
|        520 |        1 |
|        521 |       16 |
|        521 |        1 |
+------------+----------+

I've checked `assoc_type > 517 and revision > 1` and I think that `assoc_type = 517` are the review files. If so, there are only few (in our case). I cannot find too much information about the `assoc_type`'s online. Could you confirm that these are the files "_uploaded by a reviewer_" in OJS v2?

**Comment by bozana** [comment]
@retostauffer, the assoc_type = 517 is ASSOC_TYPE_REVIEW_ASSIGNMENT, s. https://github.com/pkp/pkp-lib/blob/omp-3_1_1-2/classes/core/PKPApplication.inc.php#L28. The other assoc types are also defined in that pkp-lib class and in the appropriate OJS child class (classes/core/Application.inc.php).
The corresponding assoc_id is then the review_id (that file belongs to).

**Comment by pkel** [comment]
Yes we changed the names to the original file name. But I cannot follow how `8-60-1-RV.docx` and `8-60-2-RV.pdf` imply that a user just revised a file. The naming scheme could equally be `submission-review-filenumber.extension`.

The average reviewer does at most a view reviews via OJS per year. He does not bother with
file naming schemes. He does a last minute review and uploads two files, maybe with different target audience. How should he know that he just submitted one file with two revisions? The OJS2 interface treats the uploads as separate files. That's why I think they should be separate files after upgrade.

The problem is the following.  In the old system,  the reviewer uploads comments to the author and afterwards the review itself.  Upgrade happens. The editor looks at reviewer uploads and sees multiple revisions of the same file. He was never confronted with versioned reviewer files before. He picks the last revision for inspection. Since the last revision actually contains the review, the editor will never bother with the first revision. Thereby the comments to the author are lost. This effect will get worse, the more you highlight that multiple revisions are(and have been) possible. The clearer your new design for revisions will be, the more confusion will be caused for these migrated files.

What kind of *other issues* may arise when migrating files that where displayed as separate files to files that actually are separate  files? Please help me to understand the downsides of my approach.

**Comment by bozana** [comment]
Hmmmm.... yes... 
Depending on how the issue https://github.com/pkp/pkp-lib/issues/2130 is solved, your case could not be a problem any more, once that issue is solved -- i.e. the editor would see that other file (revision) and could act appropriately, right?
But, I think the problem could be that in OJS 2 you could chose which revision of the reviewer file to show to the author and which not (regulated via viewable column). This would eventually have to be extra considered in the issue 2130 for reviewer files, although it might not be a problem because the viewable column is per revision.
From my perspective as developer who would then need to remove those reviewer revisions from the DB, insert them again in the DB as separate files, copy/rename them then appropriately in the files folder, and then consider and correct all the other dependencies seem like a problem and easy source of errors. That's why I hesitate to do so. Adding to that that the current situation seems not to be such a frequent case or at least not leading to such a big problems -- because we haven't heard it so yet --  one rather hesitate to change it. -- Just explanation of my point of view...
But actually, I/we do need to understand the problem, in order to decide -- that's why I am asking and discussing so much... also for others to have possibly clearly described everything relevant to consider here when making a decision...

@asmecher, what do you think?

We have a similar situation with revisions -- displayed separately in OJS 2 -- but they are maybe less a problem, because in OJS 3 revisions are still possible and once we display them in the issue 2130, it should not be a problem, I believe.
Thus, here is slightly different situation -- that we do not have revisions for reviewer files any more...

**Comment by asmecher** [comment]
Sorry for the delay in responding. My feeling is that we don't need something as fine-grained as per-revision access controls. Generally speaking, the Editor should be able to designate whether a file is viewable or not, and that should apply to the most recent revision. If e.g. the editor uploads a copy of the document that mistakenly exposes the reviewer's identity to the author, or vice versa, then they would resolve that by uploading a newer revision. The Reviewer and Author would not need to be able to read interim revisions that the Editor made.

**Comment by bdmckay** [comment]
@asmecher : It seems there is another case where a file uploaded by a reviewer disappears from view in the upgrade.  If a reviewer in 2.4.8 uploaded a review file but didn't select a recommendation, that file is not visible in 3.1.1.4 after upgrade.

**Comment by NateWr** [comment]
Closing this as we are no longer providing fixes for 2.x.
